jgrapht-core/src/main/java/org/jgrapht/alg/color/ColorRefinementAlgorithm.java

```diff
@@ -0,0 +1,338 @@
+     *  @param s the current color
+     * @param currentMaxColorDegree maximum color degree of s
+     * @param minColorDegree the mapping from color to its minimum color degree
+     * @param refineStack the stack refineStack that administrates all colors tobe refined
+     * @param k the number of currently used colors
+     * @param numColorDegree mapping from the color degree to the number of vertices with that color degree
+     * @param f mapping from color degrees that occur in S to newly introduced colors or to color s
+     * @param isSInStack contains whether color s is already in stack
+     * @param b index with the maximum number of vertices with the corresponding color degree
+     */
+    private void addColorsToRefineStackAndComputeF(Integer s, Integer currentMaxColorDegree, int[] minColorDegree, Deque<Integer> refineStack, Integer k, Map<Integer, Integer> numColorDegree, Map<Integer, Integer> f, boolean isSInStack, int b) {
+        for(int i = 0; i <= currentMaxColorDegree; ++i) { // go through all indices (color degrees) of numColorDegree
+            if(numColorDegree.get(i) >= 1) { // if there is a vertex with color degree i
+                if(i == minColorDegree[s]) { // i is the minimum color degree of s
+                    f.put(i, s); // colors with minimum color degree keep color s
+                    if(!isSInStack && b != i) { // push s on the stack if it is not in the stack and i is not the index with the maximum number of vertices with the corresponding color degree
+                        refineStack.push(f.get(i));
+                    }
+                } else { // i is not the minimum color degree of s
+                    k++; // we add a new color so we have to increase the number of colors
+                    f.put(i, k); // we add the new color to f for the index i
+                    if(isSInStack && i != b) { // push s on the stack if it is in the stack and i is not the index with the maximum number of vertices with the corresponding color degree
+                        refineStack.push(f.get(i));
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * checks whether alpha is a valid surjective l-coloring for the given graph
+     *
+     * @param alpha the surjective l-coloring to be checked
+     * @param graph the graph that is colored by alpha
+     * @return whether alpha is a valid surjective l-coloring for the given graph
+     */
+    private boolean isAlphaConsistent(Coloring<V> alpha, Graph<V, E> graph) {
+        // check if every vertex is assigned a color
+        if(alpha.getColors().size() != graph.vertexSet().size()) {
+            for(V v : alpha.getColors().keySet()) {
+                // TODO: What does this do? Maybe check surjectivity of alpha?
+                if(alpha.getColors().get(v) <= alpha.getNumberColors() && alpha.getColors().get(v) >= 1) {
```

Comment:
`alpha.getColors().get(v)` can be stored in a variable

Reply:
#wearedumb... lol. I fixed that